wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt

```diff
@@ -179,11 +186,18 @@ private fun parseEnum(path: List<Int>, helper: SourceCodeHelper, enum: EnumDescr
 
 private fun parseMessage(path: List<Int>, helper: SourceCodeHelper, message: DescriptorProto, descs: DescriptorSource): MessageElement {
   val info = helper.getLocation(path)
-
   val nestedTypes = mutableListOf<TypeElement>()
   val nestedMessagePath = mutableListOf(*path.toTypedArray())
   nestedMessagePath.addAll(listOf(DescriptorProto.NESTED_TYPE_FIELD_NUMBER, 0))
+
+  val mapTypes = mutableMapOf<String, MapEntryType>()
   for ((index, nestedType) in message.nestedTypeList.withIndex()) {
+    if (nestedType.options.mapEntry) {
+      val key = nestedType.fieldList[0]
+      val value = nestedType.fieldList[1]
+      mapTypes["${value.typeName}.${nestedType.name}"] = MapEntryType(key, value)
```

Comment:
I think you need this string to be `"." + fullyQualifiedEnclosingTypeName + "." + nestedTypeName`.

That way it _exactly_ matches the type name when you are processing fields. Otherwise, using `contains` or "fuzzy" matching can lead to the wrong behavior in something like this:

```protobuf
package foo.bar;

message Foo {
  // results in type name ".foo.bar.Foo.DataEntry"
  map<string, string> data = 1;
  message Bar {
    message Foo {
      message DataEntry {
      }
    }
  }
  // results in type name ".foo.bar.Foo.Bar.Foo.DataEntry"
  // since this contains "Foo.DataEntry" it will be incorrectly
  // assumed to be a map type, too
  Bar.Foo.DataEntry more_data = 2;
}
```

You may need to supply a name prefix to `parseMessages`. When parsing files this will be empty for package-less files and `package + "."` otherwise. That way, you can always compute the current message's fully-qualified name as `prefix + typeName`. Then, as you recurse to process nested messages, you supply `typeFullName + "."` as the prefix for the nested types.